### DISCUSSION TITLE
How to use Renamer to rename methods and properties in a referenced third-party library?

### DISCUSSION THREAD

**Discussion by 289908342** [dotnet/roslyn#73910] [Q&A]
Hi, I need to rename the methods and properties used in my solution that come from a third-party library. Because the methods and properties exposed by the third-party library are not always stable and may undergo disruptive or incompatible changes.

Here is my approach:
1. Scan the methods and properties used from the third-party library within the solution and collect their symbol.
2. Determine if any modifications are necessary and if they should be renamed to new names.
3. Use Renamer.RenameSymbolAsync to achieve the modifications.

In step 1, I use `syntaxRoot.DescendandNodes().OfType<IndentifierNameSyntax>()` to find them and collect their symbol, here's my code:

```csharp
var solutionPath = @"My\Test\Solution\Path";
var workspace = MSBuildWorkspace.Create();
var solution = await workspace.OpenSolutionAsync(solutionPath);

foreach (var project in solution.Projects)
{
	var compilation = await project.GetCompilationAsync();
	if (compilation == null) continue;
	
	foreach (var document in project.Documents)
	{
		var syntaxRoot = await document.GetSyntaxRootAsync();
		var sematicModel = compilation.GetSematicModel(syntaxRoot.SyntaxTree);
		var identifierNameSyntaxes = syntaxRoot.DescendantNodes().OfType<IdentifierNameSyntax>();
		foreach (var identifierName in identifierNameSyntaxes)
		{
			// Collect some IdentifierNameSyntaxes and their symbol
		}
	}
}
```

In step 2, I use SQLite to read all the names of the APIs before and after modification. But this doesn't really affect the outcome, we can simply understand that based on the original name, a new name is generated.

However, I am encountering an exception when executing step 3, Here's my code:

```csharp
var symbol = sematicModel.GetSymboInfo(identifierName).Symbol;

// System.ArgumentException: "符号 xx 不是来自源"
// In Microsoft.CodeAnalysis.Rename.Renamer.<RenameSymbolAsync>d_3.MoveNext()
await Renamer.RenameSymbolAsync(solution, symbol, "newName", solution.Options);
```

Based on my understanding, this should indicate that the symbol corresponding to IdentifierNameSyntax is not declared in the solution, hence it cannot be modified.

I am new to Roslyn, so the above ideas might be wrong. If that is the case, please forgive me and tell me what the correct approach should be, such as using regex replacement to make modifications?

**Comment by huoyaoyuan** [comment]
> Hi, I need to rename the methods and properties used in my solution that come from a third-party library.

You cannot rename the original definition of the symbol, because it's external.

> Because the methods and properties exposed by the third-party library are not always stable and may undergo disruptive or incompatible changes.

In this case, you should rename the *identifier at call sites* instead.

You may manually replicate of the implementation of renamer, and keep the rename of callers only.
